iravelmakina/simple-text-editor
# Text and File Encryption

### Description:
- Implemented `Text` class to store current text.
- Implemented `Cursor` class for managing cursor position in the text.
- Implemented `PathValidator` class for validating file paths.
- Implemented `CaesarCipher` class to handle encryption and decryption using a dynamic library.
- Enhanced `TextManager` class to utilize new classes (`Text`, `Cursor`, `PathValidator`, `CaesarCipher`).
- Added input validation to ensure different file paths for input and output files.
- Added chunk-based file processing to handle large files efficiently.


### Changes:
- `Text` class: Implemented to store text with methods for appending lines, clearing text, and printing text.
- `Cursor` class: Implemented to manage cursor position with methods for getting and setting position.
- `PathValidator` class: Implemented to validate input and output file paths.
- `CaesarCipher` class: Implemented to load encryption and decryption functions from a DLL and process files using chunks.
- `TextManager` class: Refactored to use new classes and handle text operations, file encryption, and decryption.
- Added input validation in `getUserPath()` to ensure different file paths for input and output.
- Implemented chunk-based logic in `Caesaripher` to read and write files in chunks.

### How to Test:
1. Compile the program.
2. Run the program.
3. Test the following scenarios:
    - Enter input file path, output file path (different from input), and shift value for encryption.
    - Verify the output file contains encrypted content.
    - Enter input file path (previously generated encrypted file), output file path (different from input), and shift value for decryption.
    - Verify the output file contains decrypted content matching the original file.
4. Test edge cases such as empty input file, large input file, and invalid file paths, etc.

### Screenshots:
1-2.
![1](https://github.com/iravelmakina/simple-text-editor-ivelmakina/assets/150923968/556ce253-b01c-48ea-afae-6fe032e92c51)
![2](https://github.com/iravelmakina/simple-text-editor-ivelmakina/assets/150923968/73dab610-aa68-4a48-b4c9-df1bebe0eda8)
![3](https://github.com/iravelmakina/simple-text-editor-ivelmakina/assets/150923968/fe5e825f-b35a-4cdf-8e72-3fd64b7ef029)
![4](https://github.com/iravelmakina/simple-text-editor-ivelmakina/assets/150923968/578e0f86-4824-4891-bfa4-cb2f10b179b6)
![5](https://github.com/iravelmakina/simple-text-editor-ivelmakina/assets/150923968/f244a14c-e671-46a9-864d-bbf4813083d8)
![6](https://github.com/iravelmakina/simple-text-editor-ivelmakina/assets/150923968/4cdcfe70-63cb-433a-882d-e78649c2db18)
![7](https://github.com/iravelmakina/simple-text-editor-ivelmakina/assets/150923968/7f1f12c8-41a7-4455-b5be-f8e00dc835f1)

3-4.
- Encrypt & Decrypt a Small File
![Screenshot 2024-06-28 172734](https://github.com/iravelmakina/simple-text-editor-ivelmakina/assets/150923968/427c65c5-1b5c-4e28-a0bb-a6c9f1b896a4)
![Screenshot 2024-06-28 173134](https://github.com/iravelmakina/simple-text-editor-ivelmakina/assets/150923968/3b37f0f4-0a5c-478e-8fc0-83e797a869d7)
- Encrypt & Decrypt a Large File
![Screenshot 2024-06-28 173345](https://github.com/iravelmakina/simple-text-editor-ivelmakina/assets/150923968/d0d56049-d061-4281-abe3-148b829d0fed)
![Screenshot 2024-06-28 173628](https://github.com/iravelmakina/simple-text-editor-ivelmakina/assets/150923968/871e7bf4-e7ef-4e17-b7c6-c3f2a6a6dadc)
- Encrypt & Decrypt an Empty File
![Screenshot 2024-06-28 173803](https://github.com/iravelmakina/simple-text-editor-ivelmakina/assets/150923968/94f6930c-8e48-4c90-8e6e-b40870ad76d3)
![Screenshot 2024-06-28 173901](https://github.com/iravelmakina/simple-text-editor-ivelmakina/assets/150923968/17702392-083d-4f12-87c5-086b679c4cfd)
- Validate an Empty Path
![Screenshot 2024-06-28 174024](https://github.com/iravelmakina/simple-text-editor-ivelmakina/assets/150923968/ca781f02-4847-45e0-826a-16593f720dd5)
- Validate a Non-Existent Input File Path
![Screenshot 2024-06-28 183121](https://github.com/iravelmakina/simple-text-editor-ivelmakina/assets/150923968/2213a289-a7ed-4787-8d2a-c6cde3f15f8b)
- Validate a Non-Existent Output File Path
![Screenshot 2024-06-28 182802](https://github.com/iravelmakina/simple-text-editor-ivelmakina/assets/150923968/c3eeba0d-4a81-425d-98e1-8eb8c8db6af3)
- Validate Path with Invalid Characters
![Screenshot 2024-06-28 183935](https://github.com/iravelmakina/simple-text-editor-ivelmakina/assets/150923968/f393649b-5fc9-4646-bfd8-f8b6a10b2ada)
- Validate Path longer than allowed
![Screenshot 2024-06-28 182938](https://github.com/iravelmakina/simple-text-editor-ivelmakina/assets/150923968/b5e34b92-9bcb-46f5-82ec-6d175d8313a7)
- Validate Output File Path Same as Input File Path
![Screenshot 2024-06-28 211849](https://github.com/iravelmakina/simple-text-editor-ivelmakina/assets/150923968/cbfd3ef7-15da-4dc5-b31d-3832730ee7e9)
